lib/file-source/src/file_watcher.rs

```diff
@@ -118,56 +107,21 @@ impl FileWatcher {
     pub fn read_line(&mut self, mut buffer: &mut Vec<u8>, max_size: usize) -> io::Result<usize> {
         //ensure buffer is re-initialized
         buffer.clear();
-        if let Some(ref mut reader) = self.reader {
-            // Every read we detect the current_size of the file and compare
-            // against the previous_size. There are three cases to consider:
-            //
-            //  * current_size > previous_size
-            //  * current_size == previous_size
-            //  * current_size < previous_size
-            //
-            // In the last case we must consider that the file has been
-            // truncated and we can no longer trust our seek position
-            // in-file. We MUST seek back to position 0. This is the _simplest_
-            // case to handle.
-            //
-            // Consider the equality case. It's possible that NO WRITES have
-            // come into the file _or_ that the file has been truncated and
-            // coincidentally the new writes exactly match the byte size of the
-            // previous writes. THESE WRITES WILL BE LOST.
-            //
-            // Now the greater than inequality. All of the equality
-            // considerations hold for this case. Also, consider if a write
-            // straddles the line between previous_size and current_size. Then
-            // we will be UNABLE to determine the proper start index of this
-            // write and we WILL return a partial write of length
-            // absolute_write_idx - previous_size.
-            let current_size = reader.get_ref().metadata().unwrap().size();
-            if self.previous_size <= current_size {
-                self.previous_size = current_size;
```

Comment:
Do we not need to worry about checking for truncation anymore? It's not clear to me why that would be the case.

Reply:
The only kind of truncation that could still occur is if a file is shortened. But here 'truncation' really means truncating to zero size. The regrown file should have a different fingerprint and thus get a different watcher.

Reply:
I see! That seems to answer quite a few of my comments :smile: 

If I'm understanding correctly, the strategy of using content CRC fingerprints instead of inodes means that we no longer need to worry as much about the case where a file is truncated because it will naturally get a new fingerprint and be handled as a totally different file. Does that sound about right?

Since it seems like we can now rely on `reader` not being `None`, it'd be interesting to try to refactor it away from being an `Option` at all.

Reply:
That's right. I guess some of these changes should also have been made in a separate PR. Sorry about that.

Removing the option would be more clear. I would have to add an is_dead flag or something to mark the watchers that are scheduled for deletion. Will do.